File: api/db/models/pedido.py
"""SQLAlchemy models for Pedido and DetallePedido."""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, Float, ForeignKey, Integer, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from api.db.base import Base


class PedidoModel(Base):
    __tablename__ = "pedidos"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    codigo_pedido: Mapped[str] = mapped_column(
        String(20), unique=True, index=True, nullable=False
    )
    fecha: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    cliente_ruc: Mapped[str] = mapped_column(String(11), nullable=False, index=True)
    cliente_razon_social: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    representante: Mapped[str] = mapped_column(String(255), nullable=False, default="")
    condicion_pago: Mapped[str] = mapped_column(String(50), nullable=False, default="Contado")
    estado: Mapped[str] = mapped_column(
        String(20), nullable=False, default="Borrador", index=True
    )  # Borrador, Enviado, Aprobado, Rechazado
    subtotal: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    igv: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    total: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)

    vendedor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    aprobador_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    fecha_aprobacion: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )
    motivo_rechazo: Mapped[str | None] = mapped_column(Text, nullable=True)
    observaciones: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    # Relationships
    vendedor = relationship("UserModel", back_populates="pedidos_creados", foreign_keys=[vendedor_id])
    aprobador = relationship("UserModel", back_populates="pedidos_aprobados", foreign_keys=[aprobador_id])
    detalles = relationship(
        "DetallePedidoModel", back_populates="pedido", cascade="all, delete-orphan"
    )
    devoluciones = relationship("DevolucionModel", back_populates="pedido")
    comentarios = relationship("ComentarioPedidoModel", back_populates="pedido", cascade="all, delete-orphan")


class DetallePedidoModel(Base):
    __tablename__ = "detalles_pedido"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    pedido_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("pedidos.id", ondelete="CASCADE"), nullable=False
    )
    producto_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("productos.id"), nullable=False
    )
    cantidad: Mapped[int] = mapped_column(Integer, nullable=False)
    precio_unitario: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)
    tipo: Mapped[str] = mapped_column(
        String(20), nullable=False, default="Venta"
    )  # Venta, Bonificacion
    subtotal: Mapped[float] = mapped_column(Float, nullable=False, default=0.0)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    # Relationships
    pedido = relationship("PedidoModel", back_populates="detalles")
    producto = relationship("ProductoModel")


class ComentarioPedidoModel(Base):
    __tablename__ = "comentarios_pedido"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    pedido_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("pedidos.id", ondelete="CASCADE"), nullable=False
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=False
    )
    contenido: Mapped[str] = mapped_column(Text, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now()
    )

    pedido = relationship("PedidoModel", back_populates="comentarios")
    user = relationship("UserModel")

File: src/types/pedido.ts
export type EstadoPedido = 'Borrador' | 'Enviado' | 'Aprobado' | 'Rechazado';
export type TipoDetalle = 'Venta' | 'Bonificacion';

export interface DetallePedido {
  id: string;
  pedido_id: string;
  producto_id: string;
  producto_nombre: string;
  producto_codigo_sap: string;
  cantidad: number;
  precio_unitario: number;
  bonificacion: number;
  subtotal: number;
  tipo: TipoDetalle;
}

export interface Pedido {
  id: string;
  codigo_pedido: string;
  fecha: string;
  cliente_ruc: string;
  cliente_razon_social: string;
  representante: string;
  condicion_pago: string;
  estado: EstadoPedido;
  total: number;
  subtotal: number;
  igv: number;
  vendedor_id: string;
  vendedor_nombre: string;
  vendedor_email: string;
  detalles: DetallePedido[];
  motivo_rechazo?: string;
  aprobado_por?: string;
  fecha_aprobacion?: string;
  created_at: string;
  updated_at: string;
}

export interface CreatePedidoRequest {
  cliente_ruc: string;
  representante: string;
  condicion_pago: string;
  detalles: CreateDetallePedido[];
}

export interface CreateDetallePedido {
  producto_id: string;
  cantidad: number;
  tipo: TipoDetalle;
}

export interface ApprovePedidoRequest {
  pedido_id: string;
}

export interface RejectPedidoRequest {
  pedido_id: string;
  motivo_rechazo: string;
}

export interface PedidoEstadisticas {
  total_pedidos: number;
  pedidos_aprobados: number;
  pedidos_pendientes: number;
  pedidos_rechazados: number;
  total_ventas: number;
  porcentaje_aprobados: number;
  porcentaje_pendientes: number;
  porcentaje_rechazados: number;
  variacion_mes_anterior: number;
}

export interface TopProducto {
  codigo_sap: string;
  nombre: string;
  unidades_vendidas: number;
}

export interface ActividadReciente {
  id: string;
  tipo: 'pedido_aprobado' | 'pedido_enviado' | 'pedido_rechazado' | 'devolucion_registrada' | 'comision_calculada';
  descripcion: string;
  fecha: string;
  icono: string;
}

export interface DashboardData {
  estadisticas: PedidoEstadisticas;
  top_productos: TopProducto[];
  actividad_reciente: ActividadReciente[];
  comisiones_acumuladas: number;
  devoluciones_mes: number;
  variacion_devoluciones: number;
}
